Packages tested with YAP
-------------------------

* Preprocessing tools :
    * [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) (v0.10.1)
    * [Fastq screen](http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/) (v0.4.1)
    * [FASTX-Tool kit](http://hannonlab.cshl.edu/fastx_toolkit/) (0.0.13)
        * FASTQ/A Barcode Splitter
        * FASTQ/A Clipper
        * FASTQ Quality Filter
        * FASTQ Quality Trimmer
    * Aligner algorithms :
    * [Burrows-Wheeler Aligner(BWA)](http://bio-bwa.sourceforge.net/) (0.7.4)
    * [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml) (1.0.0)
    * [Tophat](http://ccb.jhu.edu/software/tophat/index.shtml) (2.0.8b) - 2.0.8b requires boost 1.52
    * [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) (2.1.0)
    * [STAR Aligner](https://code.google.com/p/rna-star/) - (2.3.0e)
* Postprocessing toolkit :
    * [Picard](http://picard.sourceforge.net/) (1.89)
        * Multiple Metrics
        * Alignment Metrics
        * Quality Score Distribution
        * Mean Quality by Cycle
        * GC Bias Metrics
        * Insert Size Metrics
        * RNA Seq Metrics
        * Mark Duplicates
        * Calculate Hs Metrics
        * Targeted Pcr Metrics
    * [HTSeq](www-huber.embl.de/users/anders/HTSeq/doc/overview.html) (0.5.3p9) - bowtie2 compatible
    * [Cufflinks](http://cufflinks.cbcb.umd.edu/) (2.1.1) - 2.1.1 requires boost (1.52)
    * Cuffdiff
    * Cuffmerge
    * Cuffcompare
    * [GATk](www.broadinstitute.org/gatk/) (1.4-24)
    * [MACS](https://pypi.python.org/pypi/MACS2) (2.0.10)
    * [Mothur](http://www.mothur.org/) (1.33.2)
